### PR TITLE
feat: Runtime output buffer optimization

### LINKED ISSUE #3275: ✨[Feature] Output buffer optimization in runtime module
**Is your feature request related to a problem? Please describe.**

Output buffer optimization in runtime module

**Describe the solution you'd like**

- Assuming that the input shape does not change frequently, output buffer is created in previous forward()
- Latency hiding by creating the tensor for next output buffer
- Potentially Cuda and CPU(preparing next output buffer) can be overlapped

**Describe alternatives you've considered**

if runtime module maintains persistent output buffers across multiple inference runs, it allows to reuse previously allocated memory for output tensors, potentially improving performance by reducing memory allocation overhead. But it can not handle live tensors from a previous invocation. Second invocation of model will overwrite output buffer of previous run.

**Additional context**

### PR DESCRIPTION
**Pull request by keehyuna**
# Description

Latency hiding by creating the output tensor for next output buffer

Fixes #3275 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by narendasan** [comment]
@keehyuna I believe that @peri044 has some shape inference for TRT engines utility in one of his PRs can we make sure we use that or are compatible with it?

**Comment by keehyuna** [comment]
> @keehyuna I believe that @peri044 has some shape inference for TRT engines utility in one of his PRs can we make sure we use that or are compatible with it?

I think this PR doesn't have to do with fake tensor as output shape was inferred from trt function(`getTensorShape`/`get_tensor_shape`) for both c++ and python runtime.

File: core/runtime/execute_engine.cpp [hunk 1]
```diff
@@ -263,19 +284,15 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
       output_profiler_guard =
           std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->output_profile_path);
     }
+    if ((false == compiled_engine->use_pre_allocated_outputs) || shape_changed) {
```
Comment (peri044) [inline core/runtime/execute_engine.cpp:287]: !compiled_engine->use_pre_allocated_outputs ?

File: core/runtime/execute_engine.cpp [hunk 2]
```diff
@@ -83,15 +82,32 @@ bool _cudagraphs_validate_shapes(std::vector<at::Tensor> inputs, c10::intrusive_
...
-    compiled_engine->cudagraph.reset();
-    return false;
+    return true;
   }
 
-  return true;
+  return false;
+}
+
+std::vector<at::Tensor> create_output_tensors(c10::intrusive_ptr<TRTEngine> compiled_engine) {
```
Comment (peri044) [inline core/runtime/execute_engine.cpp:95]: Can we functionalize inputs allocation/creation in the execute engine similar to this ? ( I posted a similar comment in your wrapper module PR)

**Review by narendasan — COMMENTED** [review]
Create a context manager to enable this across subgraphs

File: tests/py/dynamo/runtime/test_002_cudagraphs_cpp.py [hunk 3]
```diff
@@ -17,6 +17,9 @@
     "Torch-TensorRT runtime is not available",
 )
 class TestCudagraphsCPP(TestCase):
+    def tearDown(self):
+        # Reset to default cuda graph mode after each test
+        torch_tensorrt.runtime.set_cudagraphs_mode(False)
```
Comment (keehyuna) [inline tests/py/dynamo/runtime/test_002_cudagraphs_cpp.py:22]: if multiple tests are running by pytest, other test can run on cuda graph mode. Ensure to turn off cuda graph mode after test.

File: core/runtime/TRTEngine.h [hunk 4]
```diff
@@ -30,6 +30,37 @@ using FlattenedState = std::tuple<
...
+  RuntimeStates validate_states(bool cudagraphs_enabled, bool pre_allocated_outputs_enabled, bool shape_changed) {
+    bool need_cudagraphs_record = false;
+    bool can_use_pre_allocated_outputs = false;
+
+    // Cudagraphs record is required if cudagraphs_enabled is switched to True regardless of shape change
+    if (cudagraphs_enabled && (!prev_cudagraphs_enabled || shape_changed)) {
+      need_cudagraphs_record = true;
+    }
+    // Pre-allocated output can be used when previous and current state are true without shape change
+    if (prev_pre_allocated_outputs_enabled && pre_allocated_outputs_enabled && !shape_changed) {
```
Comment (peri044) [inline core/runtime/TRTEngine.h:53]: when would `prev_pre_allocated_outputs_enabled` be true ? Is the step 2 understanding correct in the below snippet? 

```py
# step 1. This has no pre-allocation so pre_allocated_outputs_enabled and prev_pre_allocated_outputs_enabled are false 
optimized_model = torch_tensorrt.compile(
    model,
    ir="dynamo",
    enabled_precisions={torch.half},
    inputs=inputs,
)

# step 2 : enable_pre_allocated_outputs is True, so pre_allocated_outputs_enabled is True 
# But prev_pre_allocated_outputs_enabled is still false right for step 2 ? 
with torch_tensorrt.runtime.enable_pre_allocated_outputs(optimized_model):
    out_trt = optimized_model(*inputs)
```

**Review by narendasan — APPROVED** [review]
LGTM, thanks for this

**Review by peri044 — APPROVED** [review]
LGTM

**Pull request merged**